[Feature] XCM-Emulator (#2447)
paritytech/cumulus: xcm/xcm-emulator/README.md

# xcm-emulator

XCM-Emulator is a tool to emulate XCM program execution using
pre-configured runtimes, including those used to run on live
networks, such as Kusama, Polkadot, Statemine et cetera.
This allows for testing cross-chain message passing, verifying

> outcomes, weights and side-effects. It is faster than spinning up

Reviewer: It does not test the 'physical' transport of the messages as that is substituted, but the logical transportation of the messages is tested - I.e. the routing as well as the execution of the messages in the context of the foreign chain is tested.

a zombienet and as all the chains are in one process debugging using Clion is easy.

## Limitations

As the channels are mocked, using xcm-emulator tests to test

> channel setup would not be appropriate.

Author: It’s using `*Sink` to facilitate message passing as opposed to `xcm-simulator` that is using mocks. Opening of channels is not required though.

> channel setup would not be appropriate.

Reviewer 2: > Opening of channels is not required though.

what channels? e.g. hrmp channel needs to be opened, else parachain's XcmRouter->XcmpQueue sending messages wont work.

If "As the channels are mocked" mean ump/dmp/hrmp, I wouldnt say they are mocked. Just to read/write mesasges from these queues is orchestrated directly here with some thread_local queues without `relay-chain-interface / relay-chain-inprocess-interface` code, so lets say it uses testing/simplified transport layer between relaychain and parachain.

> channel setup would not be appropriate.

Author: > Could rewrite as: "As the messages do not physically go through the same messaging infrastructure there is some code that is not being tested compared to using slower E2E tests. In future it may be possible to run these XCM emulated tests as E2E tests (without changes)."

Sounds perfect. It's indeed not 100% the same, but not fully mocked either.

> channel setup would not be appropriate.

Author: Fixed [here](https://github.com/paritytech/cumulus/pull/2447)

## Alternatives

If you just wish to test execution of various XCM instructions
against the XCM VM then the `xcm-simulator` (in the polkadot
repo) is the perfect tool for this.

## How to use

> 1. Setup a relay chain:

Reviewer 3: ```suggestion
1. Set up a relay chain:
```

```rust
decl_test_relay_chain! {
	pub struct KusamaRelay {
		Runtime = kusama_runtime::Runtime,
		XcmConfig = kusama_runtime::xcm_config::XcmConfig,
		new_ext = kusama_ext(),
	}
}
```

2. Define some parachains:
```rust
decl_test_parachain! {
	pub struct ParachainA {
		Runtime = test_runtime::Runtime,
		RuntimeOrigin = test_runtime::RuntimeOrigin,
		XcmpMessageHandler = test_runtime::XcmpQueue,
		DmpMessageHandler = test_runtime::DmpQueue,
		new_ext = parachain_ext(1),
	}
}
```

> 3. Setup a network with all the declared parachains and a relay chain:

Reviewer 3: ```suggestion
3. Set up a network with all the declared parachains and a relay chain:
```

# xcm-emulator

xcm-emulator uses relay chain and parachain runtimes.
Users could plug in Kusama, Statemine, Polkadot and other pre-configured runtimes.

> With up-to-date chain specs, it's able to verify if specific XCM messages work in live networks.

Reviewer 3: Not OK for documentation.

Reviewer: Extended the documentation to point out that xcm-simulator also exists and when to use one over the other.

# xcm-emulator

> XCM-Emulator is a tool to emulate XCM program execution using pre-configured runtimes, including those
> used to run on live networks, such as Kusama, Polkadot, Statemine et cetera.  
> This allows for testing cross-chain message passing, verifying outcomes, weights and side-effects.

Reviewer: ```suggestion
XCM-Emulator is a tool to emulate XCM program execution using
pre-configured runtimes, including those used to run on live
networks, such as Kusama, Polkadot, Statemine et cetera.
This allows for testing cross-chain message passing, verifying
outcomes, weights and side-effects. It is faster than spinning up
a zombienet and as all the chains are in one process debugging using Clion is easy.

## Limitations

As the channels are mocked, using xcm-emulator tests to test
channel setup would not be appropriate.

## Alternatives

If you just wish to test execution of various XCM instructions
against the XCM VM then the `xcm-simulator` (in the polkadot
repo) is the perfect tool for this.
```
